lib/types.nix

```diff
@@ -710,6 +710,23 @@ rec {
         in mergedOption.type;
     };
 
+    # A type that is one of several submodules, similiar to types.oneOf but is usable inside attrsOf or listOf
+    # submodules need an option with a type str which is used to find the corresponding type
+    taggedSubmodules =
+      { types
+      , specialArgs ? {}
+      }: mkOptionType rec {
+      name = "taggedSubmodules";
+      description = "one of ${concatStringsSep "," (attrNames types)}";
+      check = x: if x ? type then types.${x.type}.check x else throw "No type option set in:\n${lib.generators.toPretty {} x}";
```

Comment:
`check` should return a bool, as it's used by `either` to do rudimentary but lazy type checks.

Assuming we keep the syntax as `{ type = "gpt"; partitions = ...; }`, this should do an `evalModules` to get `type` option value, with something like `freeformType = lazyAttrsOf raw;` to ignore the other options. Alternatively we could handle the module arguments in a custom manner if that leads to better error messages.